Frameworks/Starboard/EbrStorageFile.cpp

```diff
@@ -0,0 +1,351 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include "EbrStorageFile.h"
+#include "ErrorHandling.h"
+
+#include "StringHelpers.h"
+#include "WRLHelpers.h"
+
+#include "RawBuffer.h"
+
+#include <vector>
+#include <fcntl.h>
+
+using namespace Microsoft::WRL;
+using namespace Windows::Foundation;
+using namespace WRLHelpers;
+using namespace ABI::Windows::Storage;
+using namespace ABI::Windows::Storage::AccessCache;
+using namespace ABI::Windows::Storage::FileProperties;
+using namespace ABI::Windows::Storage::Streams;
+
+std::shared_ptr<EbrFile> EbrStorageFile::CreateInstance(const char* path, int mode, int share, int pmode) {
+    // Check to see if this path is a file id path.
+    static const std::string c_fileIdPreamble = "/.file/id=";
+
+    RETURN_NULL_IF_FALSE(0 == strncmp(path, c_fileIdPreamble.c_str() , c_fileIdPreamble.length()));
+
+    Wrappers::HString idHString = Strings::NarrowToWide<HSTRING>(&path[c_fileIdPreamble.length()]);
+
+    ComPtr<IStorageApplicationPermissionsStatics> storageApplicationPermissionsStatics;
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // Try to grab the item based on id.
+    ComPtr<IStorageFile> storageFile;
+    RETURN_NULL_IF_FAILED(AwaitResult(accessList, &IStorageItemAccessList::GetFileAsync, idHString.Get(), &storageFile));
+
+    bool shouldAllowWrite = ((mode & _O_WRONLY) != 0) ||
+                            ((mode & _O_RDWR) != 0);
+    
+    FileAccessMode openMode = shouldAllowWrite ? FileAccessMode_ReadWrite : FileAccessMode_Read;
+
+    ComPtr<IRandomAccessStream> randomAccessStream;
+    RETURN_NULL_IF_FAILED(AwaitResult(storageFile, &IStorageFile::OpenAsync, openMode, &randomAccessStream));
+
+    bool shouldAppend = ((mode & _O_APPEND) != 0);
+
+    return std::make_shared<EbrStorageFile>(
+        std::move(storageFile),
+        std::move(randomAccessStream),
+        shouldAppend,
+        shouldAllowWrite);
+}
+
+EbrStorageFile::EbrStorageFile(
+        ComPtr<IStorageFile>&& storageFile, 
+        ComPtr<IRandomAccessStream>&& randomAccessStream,
+        bool shouldAppend,
+        bool shouldAllowWrite) : 
+            m_storageFile(storageFile), 
```

Comment:
>storageFile [](start = 26, length = 11)

Why pass-by-rvalue-reference with std::move on the outside (keeping external consumers aware of our internals) instead of pass-by-value with std::move on the inside? There does not appear to be consensus on this on the internet at large, but it looks like a lot of people lean towards the second.